backend/src/main/kotlin/app/ehrenamtskarte/backend/verification/Argon2IdHasher.kt

```diff
@@ -0,0 +1,76 @@
+import org.bouncycastle.crypto.params.Argon2Parameters
+import java.nio.charset.StandardCharsets
+import java.util.Base64
+
+class Argon2IdHasher {
+    companion object {
+        /**
+         * Copied from spring-security Argon2EncodingUtils.java licenced under Apache 2.0
+         *
+         * Encodes a raw Argon2-hash and its parameters into the standard Argon2-hash-string
+         * as specified in the reference implementation
+         * (https://github.com/P-H-C/phc-winner-argon2/blob/master/src/encoding.c#L244):
+         *
+         * {@code $argon2<T>[$v=<num>]$m=<num>,t=<num>,p=<num>$<bin>$<bin>}
+         **/
+        @Throws(IllegalArgumentException::class)
+        fun encode(
+            hash: ByteArray?,
+            parameters: Argon2Parameters
+        ): String? {
+            val b64encoder: Base64.Encoder = Base64.getEncoder().withoutPadding()
+            val stringBuilder = StringBuilder()
+            val type =
+                when (parameters.type) {
+                    Argon2Parameters.ARGON2_d -> "\$argon2d"
+                    Argon2Parameters.ARGON2_i -> "\$argon2i"
+                    Argon2Parameters.ARGON2_id -> "\$argon2id"
+                    else -> throw IllegalArgumentException("Invalid algorithm type: " + parameters.type)
+                }
+            stringBuilder.append(type)
+            stringBuilder
+                .append("\$v=")
+                .append(parameters.version)
+                .append("\$m=")
+                .append(parameters.memory)
+                .append(",t=")
+                .append(parameters.iterations)
+                .append(",p=")
+                .append(parameters.lanes)
+            if (parameters.salt != null) {
+                stringBuilder.append("$").append(b64encoder.encodeToString(parameters.salt))
+            }
+            stringBuilder.append("$").append(b64encoder.encodeToString(hash))
+            return stringBuilder.toString()
+        }
+
+        fun hashKoblenzUserData(userData: KoblenzUser): String? {
+            val canonicalJson = CanonicalJson.koblenzUserToString(userData)
+            val hashLength = 32
+
+            val pepper = Environment.getVariable(KOBLENZ_PEPPER_SYS_ENV) // TODO handle if Null
+            val pepperByteArray = pepper?.toByteArray(StandardCharsets.UTF_8)
+            val params =
+                Argon2Parameters
+                    .Builder(Argon2Parameters.ARGON2_id)
+                    .withVersion(19)
+                    .withIterations(2)
+                    .withSalt(pepperByteArray)
+                    .withParallelism(1)
+                    .withMemoryAsKB(16)
+                    .build()
+
+            val generator = Argon2BytesGenerator()
+            generator.init(params)
+            val result = ByteArray(hashLength)
+            generator.generateBytes(canonicalJson.toCharArray(), result)
+            return encode(result, params)
```

Comment:
I think we should not use the encoding that you suggest (instead, we can simply use the base64 encoding of `result`).
There's simply no reason to add these parameters to each hash as we will use the hash as an index and not as a password (So we will have to stick to these parameters unless we can recompute all hashes -- which would need to be done by Koblenz). Also I think, we use the salt as a global credential. Then, it does not make sense to append it to every hash.

Reply:
ah i think we talked about some kind of "minimal" validation. But i agree that the that the encoded form is not very handy.
What do you mean with "append". We use here the salt as a pepper (global) to even make it more difficult to brute force it

Reply:
The `encode` function adds the parameters as well as the salt to the "encoded" hash. I suggest to not use the `encode` function and instead simply return `base64Encode(result)`.

Reply:
Hmh, I think we should keep the encoding (it also is quite a standart, not something I invented), as we can do some minimal form of validation and testing, when something is not working as expected. At least we can check for parameters being set and we will not get confused if we change something in the future, e.g. the salt or the version. It can be really hard (or impossible) without this encoding to check what is going wrong, so I would really prefer to keep this encoding, as the only downside of keeping it, is slightly longer entry in the database, but I think this is not really important for the 15-20k entries.

Reply:
We should probably at least strip the secret salt from the hashes when writing the hashes to the DB. That's the only problem I see with the encoding.